src/index.spec.ts

```diff
@@ -89,183 +88,310 @@ describe('EppoClient E2E test', () => {
+      const initPromise = init({
+        apiKey: 'dummy',
+        baseUrl: `http://127.0.0.1:${TEST_SERVER_PORT}`,
+        assignmentLogger: mockLogger,
+      });
+
+      // Advance timers mid-init to allow retrying
+      await jest.advanceTimersByTimeAsync(maxRetryDelay);
+
+      // Await so it can finish its initialization before this test proceeds
+      await initPromise;
+
+      const client = getInstance();
+      expect(client.getStringAssignment('subject', flagKey)).toBe('control');
     });
-    const client = new EppoClient(mockConfigRequestor, mockPoller);
-    const assignment = client.getAssignment('subject-10', flagKey);
-    expect(assignment).toEqual('variant-2');
-  });
 
-  it('returns null when experiment config is absent', () => {
-    const mockConfigRequestor = td.object<ExperimentConfigurationRequestor>();
-    const mockPoller = td.object<IPoller>();
-    td.when(mockConfigRequestor.configStore.get(flagKey)).thenReturn(null);
-    const client = new EppoClient(mockConfigRequestor, mockPoller);
-    const assignment = client.getAssignment('subject-10', flagKey);
-    expect(assignment).toEqual(null);
-  });
+    it('gives up initial request and throws error after hitting max retries', async () => {
+      td.replace(HttpClient.prototype, 'get');
+      let callCount = 0;
+      td.when(HttpClient.prototype.get(td.matchers.anything())).thenDo(async () => {
+        callCount += 1;
+        return Promise.reject(new Error('Intentional Thrown Error For Test'));
+      });
 
-  it('logs variation assignment and experiment key', () => {
-    const mockConfigRequestor = td.object<ExperimentConfigurationRequestor>();
-    const mockLogger = td.object<IAssignmentLogger>();
-    const mockPoller = td.object<IPoller>();
-    td.when(mockConfigRequestor.configStore.get(flagKey)).thenReturn(mockExperimentConfig);
-    const subjectAttributes = { foo: 3 };
-    const client = new EppoClient(mockConfigRequestor, mockPoller);
-    client.setLogger(mockLogger);
-    const assignment = client.getAssignment('subject-10', flagKey, subjectAttributes);
-    expect(assignment).toEqual('control');
-    expect(td.explain(mockLogger.logAssignment).callCount).toEqual(1);
-    expect(td.explain(mockLogger.logAssignment).calls[0].args[0].subject).toEqual('subject-10');
-    expect(td.explain(mockLogger.logAssignment).calls[0].args[0].featureFlag).toEqual(flagKey);
-    expect(td.explain(mockLogger.logAssignment).calls[0].args[0].experiment).toEqual(
-      `${flagKey}-${mockExperimentConfig.rules[0].allocationKey}`,
-    );
-    expect(td.explain(mockLogger.logAssignment).calls[0].args[0].allocation).toEqual(
-      `${mockExperimentConfig.rules[0].allocationKey}`,
-    );
-  });
+      // Note: fake time does not play well with errors bubbled up after setTimeout (event loop,
+      // timeout queue, message queue stuff) so we don't allow retries when rethrowing.
```

Comment:
This cost me probably 5 hours of futile attempts to have this test include retrying once  before throwing. :rage4: 

Note that when _not_ using fake timers and just setting a very short interval, this works with retries, so I'm feeling good about it.